### PR TITLE
EgressService: Seperate Network and SNAT

### PR DESCRIPTION
**Pull request by oribon**
**- What this PR does and why is it needed**

Implement https://github.com/openshift/enhancements/pull/1382 - allowing the user to specify Network without SNATing to the LoadBalancer's ingress IP.

**- Special notes for reviewers**

E2E commit fixes https://github.com/ovn-org/ovn-kubernetes/issues/3584
Last commit ("sync node labels") isn't really related to the CRD change, but added it here since it's pretty small and not worth wasting more ci cycles in another PR, resolves https://github.com/ovn-org/ovn-kubernetes/pull/3389#discussion_r1175231317

**- How to verify it**

**- Description for the changelog**

EgressService: Support separated Network and LoadBalancer SNAT.

### CODE REVIEW

File: go-controller/pkg/clustermanager/egressservice/egressservice_cluster_node.go
Comment (kyrtapz) [inline go-controller/pkg/clustermanager/egressservice/egressservice_cluster_node.go:212]: We are removing all of the egressservice labels for nodes that do not have a state in our cache, I am concerned that this could race on startup where we would remove all labels here for them to be added back in the esvc controller. 
Shouldn't we do it in the main controller instead?

**Review by kyrtapz — CHANGES_REQUESTED** [review]
Overall the changes look good.

I think in the second commit you are missing cleaning up the node labels for services that were changed from `SourceIPLoadBalancer` to `SourceIPNetwork` when ovn-k was not running.

File: go-controller/pkg/ovn/controller/egressservice/egressservice_zone.go
Comment (trozet) [reply go-controller/pkg/ovn/controller/egressservice/egressservice_zone.go:672]: we do have a state here right? otherwise we would have returned a few lines up

Comment (kyrtapz) [reply go-controller/pkg/ovn/controller/egressservice/egressservice_zone.go:672]: What I meant was that we do not normally have state stored for services that have `noSNATHost`.
This means that if we hit line 672 the service likely changed from `SourceIPLoadBalancer` to `SourceIPNetwork`

File: go-controller/pkg/clustermanager/egressservice/egressservice_cluster.go
Comment (oribon) [reply go-controller/pkg/clustermanager/egressservice/egressservice_cluster.go]: done

File: go-controller/pkg/clustermanager/egressservice/egressservice_cluster.go
Comment (oribon) [reply go-controller/pkg/clustermanager/egressservice/egressservice_cluster.go]: I modified the function to accept a new "serviceHost" enum (that represents hosts with special meanings) to set on the status instead of defaulting to "", is that approach fine?

File: go-controller/pkg/clustermanager/egressservice/egressservice_cluster.go
Comment (oribon) [reply go-controller/pkg/clustermanager/egressservice/egressservice_cluster.go]: done

File: go-controller/pkg/clustermanager/egressservice/egressservice_cluster.go
Comment (oribon) [reply go-controller/pkg/clustermanager/egressservice/egressservice_cluster.go]: nope, because I like avoiding nested ifs (:sweat_smile:) and I think it won't do anything if it's already the desired status. do you prefer I handle it anyways (with a nested if or another conditional)?

File: go-controller/pkg/ovn/controller/egressservice/egressservice_zone.go
Comment (oribon) [reply go-controller/pkg/ovn/controller/egressservice/egressservice_zone.go:674]: right, added the comment for it

File: go-controller/pkg/ovn/egressservices_test.go
Comment (oribon) [reply go-controller/pkg/ovn/egressservices_test.go]: done

File: go-controller/pkg/node/egress_service_test.go
Comment (oribon) [reply go-controller/pkg/node/egress_service_test.go]: nope, removed

File: docs/egress-service.md
Comment (trozet) [inline docs/egress-service.md:25]: typo here with "create s"

File: docs/egress-service.md
Comment (trozet) [inline docs/egress-service.md:107]: is there a check in your current code to detect this unsupported case and log an error or post an event to the egress service crd?

Comment (oribon) [reply docs/egress-service.md:107]: added a log in the clustermanager

File: docs/egress-service.md
Comment (oribon) [reply docs/egress-service.md]: fixed

File: go-controller/pkg/node/controllers/egressservice/egressservice_node.go
Comment (oribon) [reply go-controller/pkg/node/controllers/egressservice/egressservice_node.go:849]: is there a guarantee that in some point it is available? if so then it makes sense I guess

Comment (trozet) [reply go-controller/pkg/node/controllers/egressservice/egressservice_node.go:849]: the field in the crd is optional, im not sure if kube controller always sets it for automatically created endpoints. I guess it is OK.

**Comment by trozet** [comment]
@kyrtapz will wait for your lgtm before I merge

**Review by kyrtapz — APPROVED** [review]

**Pull request merged**
